### ISSUE TITLE
Feature to opt-out of force-pushing pure rebases when no merge conflict exists in the previous PR to reduce notification noise

### ISSUE THREAD

**Issue by consideRatio** [peter-evans/create-pull-request#1190]
### Subject of the issue

A feature proposal, a boolean input to the action allowing users to opt-out of (or opt-in to) force-pushing pure rebases when no merge conflict exists in the previous PR. The goal is to be able to favor less notifications over more recently rebased PRs.

In the [action behavior](https://github.com/peter-evans/create-pull-request#action-behaviour) topic of the readme, its documented that 

> - If a pull request already exists and there are no further changes (i.e. no diff with the current pull request branch) then the action exits silently.

I think whats not obvious from the note above, is that "no further changes" also includes unrelated changes made in the repo. A previously opened PR with the same changes will get force-pushed if there was an unrelated commit made in the default branch since last time. This is unexpected behavior to me, and creates regular noise that I'd like to opt-out of. I say opt-out because the default behavior of rebasing can make sense, but can also be considered not worth the notification noise.

I think the behavior of @dependabot by default is to not do anything unless there is a merge conflict detected. But, I don't understand how this is done technically yet.

### Technical implementation brainstorm

1. Is configuration to not always rebasing is set?
1. Is there is a merge conflict in the PR (hopefully a GitHub API declares this)?
1. Are there _proper differences_ in the existing branch and the new state branch?
   I think _proper differences_ is to ask if one is solely is a rebase of the other, or if there are more changes like a change to the commit message etc. I'm not sure if this is technically reasonable or not. I'm also not sure how @dependabot's logic is, but that could probably used as an inspiration.

### Steps to reproduce

To conclude that the current behavior is to force push if there is an unrelated change: let this action open a PR making one change, then push an unrelated change to the default branch, and let this action do its job like before and a force push will be made.

**Comment by peter-evans** [comment]
Hi @consideRatio 

Just to explain why the action behaves this way, please see [this comment](https://github.com/peter-evans/create-pull-request/issues/1062#issuecomment-1044548704). There are also some workarounds mentioned there that might work for your use case.

I would like to try and avoid changing the action's behaviour, as explained in the linked comment. My feeling is that use case specific decision making should ideally be in the workflow, where a decision can be made to execute the action, or not.

It sounds like the ideal behaviour in your case is to not do anything unless there is a merge conflict. I think that should be possible to detect in a separate workflow step. Then a decision can be made to execute this action, or not.

**Comment by consideRatio** [comment]
@peter-evans thank you for your time and work on this project, I think you decision makes sense for the sustainability of this project given that it can indeed complicate things significantly to do this etc.

Instead of directly closing this issue, maybe there is something that can be done about this comment?

> If a pull request already exists and there are no further changes (i.e. no diff with the current pull request branch)

I think one part that is unclear is that `no diff` part. While `git diff` would list all changes between commits, the GitHub UI would only include the "relevant changes" that was part of the PR.

**Comment by peter-evans** [comment]
> I think one part that is unclear is that no diff part. While git diff would list all changes between commits, the GitHub UI would only include the "relevant changes" that was part of the PR.

What the action does is recreate the pull request branch locally using the latest base. Then it does it `git diff` between its local version and the existing pull request branch. If there is no diff, then the pull request branch is not updated. If there is a diff the pull request branch is force pushed to update it (as if it was created new that run).

Perhaps what is lacking from the explanation if the documentation is the fact that changes on the base branch (the target) will cause the pull request to be updated. Is that what you mean about it being unclear?

**Comment by consideRatio** [comment]
> Perhaps what is lacking from the explanation in the documentation is the fact that changes on the base branch (the target) will cause the pull request to be updated. Is that what you mean about it being unclear?

Yes!

> What the action does is recreate the pull request branch locally using the latest base. Then it does agit diff between its local version and the existing pull request branch

I've reread this a few times and I think a confusion point for me was the wording "recreate the pull request branch". It sounds like the existing PR branch is used in this process somehow, but in practice it isn't.

I understand it as this action will:
- commit dirty git state (new commit)
- compare its current git state with the PR branch content wise (ignoring commit messages etc) via `git diff`
- sometimes force push even if the "new commit" is already part of the PR branch, because the `git diff` find changes if a rebase would been needed as well, or if additional commits have been pushed to the existing PR branch.

**Comment by peter-evans** [comment]
I've updated the description. Hopefully this is a little clearer. I realised that the bit about "no diff with the current pull request branch" doesn't make much sense if you don't know how the action works. It wasn't good to expect that knowledge, so thanks for pointing it out.

> If a pull request already exists it will be updated if necessary. Local changes in the Actions workspace, or changes on the base branch, can cause an update. If no update is required the action exits silently.

Your understanding of how the action works is correct.

I'll close this now, but feel free to continue if you would like to discuss anything further.
